File: checkRelease.py
import argparse
import json
import subprocess
import sys
from typing import List
import os


class Branch:
    name: str
    is_merged: bool

    def __init__(self, name) -> None:
        self.name = name
        self.is_merged = False


class AsanaTask:
    name: str
    id: int
    branches: List[Branch]
    has_reverts: bool
    url: str

    def __init__(self, name, id, url) -> None:
        self.name = name
        self.id = id
        self.url = url
        self.branches = []
        self.has_reverts = False

    def is_ready(self):
        return len(self.branches) == 1 and self.branches[0].is_merged and not self.has_reverts


def run_in_terminal(command):
    return subprocess.run(command, shell=True, capture_output=True, encoding='utf-8')


def parse_args():
    parser = argparse.ArgumentParser()

    parser.add_argument(
        "-u", "--url", type=str,
        help="URL of Asana Project. Example: -u https://app.asana.com/0/1201640226677955/list")

    parser.add_argument(
        "-s", "--sections", type=str,
        help="Sections in the project that need to be checked. Example: --sections 'ClickDeploy, Stories to Deploy'")

    parser.add_argument(
        "-t", "--token", type=str, help="You can pass Asana Access Token here, if Environment variable doesn't work. Example: --token '1/1200261239008160:0c75d3a830cfa6c7e7c6f8856cad3b21'")

    parser.add_argument(
        "--short", action="store_true", help="Show problems only (marked as !!)", default=False)

    return parser.parse_args()


def get_asana_token(hardcoded_asana_token, arg_asana_token, env_var_name):
    if hardcoded_asana_token:
        return hardcoded_asana_token
    elif arg_asana_token:
        return arg_asana_token

    try:
        return os.environ[env_var_name]
    except:
        print(f"Couldn't find any Asana Access Token in Environment variable {env_var_name}, --token Argument or Hardcoded")
        sys.exit()


def validate_token(token):
    response = run_in_terminal(f"""curl -X GET https://app.asana.com/api/1.0/users/me -H 'Authorization: Bearer {token}'""")
    if "Not Authorized" in response.stdout:
        print("Your Asana Access Token is not authorized")
        sys.exit()


def handle_help_arg(arguments):
    if '-h' in sys.argv:
        print(arguments.h)
        sys.exit()


def handle_sections_arg(arg_sections, default_sections):
    if not arg_sections:
        arg_sections = default_sections

    arg_sections_list = arg_sections.upper().split(",")
    for i, section in enumerate(arg_sections_list):
        arg_sections_list[i] = section.strip()

    return arg_sections_list


def validate_project_id(project_id):
    if len(project_id) != 16 or not project_id.isnumeric():
        print("Project ID is invalid, it must be 16 characters, numbers only")
        sys.exit()


def get_section_ids(project_id, token, arg_sections_arg: List[str]) -> dict:
    sections = run_in_terminal(f"""curl -X GET https://app.asana.com/api/1.0/projects/{project_id}/sections -H 'Accept: application/json' -H 'Authorization: Bearer {token}'""").stdout
    if "Not a recognized ID" in sections:
        print(f"Project with ID {project_id} doesn't exist")
        sys.exit()
    json_sections = json.loads(sections)

    section_ids = {}
    for section in json_sections['data']:
        if section['name'].upper() in arg_sections_arg:
            section_ids[section['name'].upper()] = section['gid']

    for section_name in arg_sections_arg:
        if not section_ids.get(section_name):
            print(f'Section "{section_name}" was not found')

    return section_ids


def get_tasks_in_json(section_ids: List[str], token):
    json_tasks = []
    for section_id in section_ids:
        tasks_raw = run_in_terminal(f"""curl -X GET https://app.asana.com/api/1.0/tasks?section={section_id} -H 'Accept: application/json' -H 'Authorization: Bearer {token}'""")
        json_tasks.append(json.loads(tasks_raw.stdout)['data'])
    return json_tasks


def get_tasks_list(tasks_json_for_sections, base_url) -> List[AsanaTask]:
    tasks: List[AsanaTask] = []
    for tasks_json in tasks_json_for_sections:
        for task in tasks_json:
            task_name = task['name']
            task_id = task['gid']
            task_url = f"{base_url}{task_id}"

            task_object = AsanaTask(task_name, task_id, task_url)
            tasks.append(task_object)

    return tasks


def handle_tasks(tasks: List[AsanaTask], short):
    print('\n')
    for task in tasks:
        # Find remote Branches
        task_branches = run_in_terminal(f"git branch --remotes | grep {task.id} | tr '\n' ' '")
        if not task_branches.stdout:
            if not short:
                print_task(task)
            continue

        # Add Branches to object
        task_branches_split = task_branches.stdout.strip().split("  ")
        for branch in task_branches_split:
            task.branches.append(Branch(branch.strip()))

        # Check for reverts
        revert_commits = run_in_terminal(f'git log --oneline | grep {task.id} | grep -i revert')
        if revert_commits.stdout != "":
            task.has_reverts = True

        # Check if fully merged
        if len(task.branches) > 0:
            for branch in task.branches:
                last_commit = run_in_terminal(f"git log {branch.name} -1 --oneline | awk '{{print $1}}'")
                is_merged = run_in_terminal(f"git log --oneline | grep {last_commit.stdout}")

                if is_merged.stdout != "":
                    branch.is_merged = True

        # Skip tasks without problems when --short
        if short and task.is_ready():
            continue
        print_task(task)


def print_task(task: AsanaTask):
    if task.has_reverts:
        print(f'!! Has Reverts')
    if len(task.branches) > 1:
        print(f'!! Found {len(task.branches)} branches')
    print(f"   {task.name} -> {task.url}")

    if len(task.branches) == 0:
        branch_print = f"No Branch"
        print(f'   {task.id} -> {branch_print}\n\n')
        return

    for branch in task.branches:
        branch_print = branch.name

        if branch.is_merged:
            ok_print = "OK "
            is_merged_print = " -> Merged"
        else:
            ok_print = "!! "
            is_merged_print = " -> NOT Merged"

        print(f'{ok_print}{task.id} -> {branch_print}{is_merged_print}')
    print("\n")


# Default variables
hardcoded_asana_token = ''
default_sections = 'ClickDeploy,Stories to Deploy'
env_var_name = "ASANA_TOKEN"

# Argument valiables
arguments = parse_args()
arg_asana_token = arguments.token
arg_short = arguments.short
arg_sections = arguments.sections
arg_project_url = arguments.url

# Asana client
asana_token = get_asana_token(hardcoded_asana_token, arg_asana_token, env_var_name)
validate_token(asana_token)

# Arguments parsing
handle_help_arg(arguments)
sections_list = handle_sections_arg(arg_sections, default_sections)
base_url = arg_project_url[:41]
project_id = arg_project_url.split("/")[-2]
validate_project_id(project_id)

# Getting info from Asana API
section_ids = get_section_ids(project_id, asana_token, sections_list)
tasks_json_for_sections = get_tasks_in_json(list(section_ids.values()), asana_token)
tasks = get_tasks_list(tasks_json_for_sections, base_url)
# tasks.append(AsanaTask('Task dummy', 1201636415789062, 'TestURL')) 
# tasks.append(AsanaTask('Task dummy 1', 1200783749941177, 'TestURL'))
handle_tasks(tasks, arg_short)

File: checkRelease/checkRelease.js
import terminal from 'child_process';
import asana from 'asana';
import spinner from 'ora';
import { Command } from 'commander'
const argParser = new Command();

//  Default variables
const defaultSections = 'ClickDeploy,Stories to Deploy,User Stories,Deployment';
const hardcodedAsanaToken = '';
const envVarName = "ASANA_TOKEN";

class ReleaseValidator {
    arguments;
    baseURL;
    projectId;
    asanaClient;

    constructor() {
        this.arguments = this.parseArguments();
        this.baseURL = this.arguments.url.substring(0,41);
        this.projectId = this.arguments.url.split("/").slice(-2)[0];
        this.validateProjectId();

        const asanaToken = this.getAsanaToken();
        this.validateToken(asanaToken);
        this.asanaClient = asana.Client.create({defaultHeaders: {'Asana-Enable': 'new_memberships,new_goal_memberships'}}).useAccessToken(asanaToken);
    }

    async run() {
        this.fetchOrigin();
        let sections = await this.getSections();
        let tasksFromSectionsJson = await this.getTasksFromSectionsJson(sections);
        let tasks = this.getListOfTaskObjects(tasksFromSectionsJson);
        this.handleBranches(tasks);
        this.printTasks(tasks);
    }

    runInTerminal(command, showOutput = false) {
        try {
            const stdioOption = showOutput ? 'inherit' : 'pipe';
            return terminal.execSync(command, { stdio: stdioOption }).toString();
        } catch (error) {
            return null;
        }
    }

    parseArguments() {
        argParser
            .requiredOption('-u, --url <value>', 'URL of Asana Project. Example: -u https://app.asana.com/0/1201640226677955/list')
            .option('-s, --sections [value]', "Sections in the project that need to be checked. Example: --sections 'ClickDeploy, Stories to Deploy'", defaultSections)
            .option('-a, --all-sections', 'Check all Sections in a Project', false)
            .option('-sp, --sprint [value]', "Filter checked tasks by Sprint. Example: --sprint 'Sprint 24'")
            .option('-t, --token [value]', "You can pass Asana Access Token here, if Environment variable doesn't work. Example: --token '1/1200261289008160:0c75d3a830cfa6c7e7c6f8856cad3b21'")
            .option('--short', 'Show problems only (marked as !!)', false)
    
        return argParser.parse(process.argv).opts();
    }
    
    getAsanaToken() {
        if (hardcodedAsanaToken) {
            return hardcodedAsanaToken;
        } else if (this.arguments.token) {
            return this.arguments.token;
        }
    
        if (process.env[envVarName]) {
            return process.env[envVarName];
        } else {
            console.log(`Couldn't find any Asana Access Token in Environment variable ${envVarName}, --token Argument or Hardcoded`);
            process.exit(1);
        }
    }
    
    validateToken(token) {
        let sp = spinner('Authorization');
        sp.start();

        // let authResult = this.runInTerminal(`curl -X GET https://app.asana.com/api/1.0/users/me -H 'Authorization: Bearer ${token}'`);
        // let authResult = terminal.execSync(`curl -X GET https://app.asana.com/api/1.0/users/me -H 'Authorization: Bearer ${token}'`, {stdio:'pipe'}).toString();
        let authResult = this.runInTerminal(`curl -X GET https://app.asana.com/api/1.0/users/me -H 'Authorization: Bearer ${token}'`);
        if(authResult.includes('Not Authorized')) {
            sp.fail('Your Asana Access Token is not authorized');
            process.exit(1);
        }
        sp.succeed();
    }
    
    validateProjectId() {
        if (this.projectId.length != 16 || !this.projectId.match(/^\d+$/)) {
            console.log('Project ID is invalid, it must be 16 characters, numbers only');
            process.exit(1);
        }
    }

    fetchOrigin() {
        let sp = spinner('Fetching Origin');
        sp.start();
        terminal.execSync(`git fetch --all`, {stdio: 'ignore'});
        sp.succeed();
    }
    
    async getSections() {
        let sp = spinner('Retrieving Sections');
        sp.start();

        let chosenSectionsStringList = []
        if (!this.arguments.allSections) {
            if (this.arguments.sections) {
                chosenSectionsStringList = this.arguments.sections.split(',').map((section) => section.trim().toUpperCase());
            }
        }
    
        return await this.asanaClient.sections.getSectionsForProject(this.projectId, {opt_pretty: true})
            .then(
                (result) => {
                    sp.succeed();
                    let jsonSections = result.data;
    
                    let retrievedSectionsObjects = [];
                    for (const section of jsonSections) {
                        if (this.arguments.allSections || chosenSectionsStringList.includes(section.name.toUpperCase())) {
                            retrievedSectionsObjects.push(new Section(section.name.toUpperCase(), section.gid));
                        }
                    }
    
                    if (this.arguments.allSections) return retrievedSectionsObjects;
    
                    // Check not found sections
                    // let retrievedSectionsNames = retrievedSectionsObjects.map((s) => s.name);
                    // for (const sectionName of chosenSectionsStringList) {
                    //     if (!retrievedSectionsNames.includes(sectionName)) {
                    //         console.log(`Section "${sectionName}" was not found`);
                    //     }
                    // }
                    return retrievedSectionsObjects;
                },
                (error) => {
                    sp.fail(`Project with ID ${this.projectId} doesn't exist`);
                    process.exit(1);
                }
            );
    }
    
    async getTasksFromSectionsJson(sections) {
        let jsonTasks = [];
        for (const section of sections) {
            let sp = spinner(`Retrieving Tasks from Section: ${section.name}`);
            sp.start();
            let params = "gid,name,custom_fields";
            await this.asanaClient.tasks.getTasksForSection(section.id, {opt_fields: `${params}`, opt_pretty: true, limit: 100})
                .then(
                    (result) => {
                        sp.succeed();
                        jsonTasks.push(result.data);
                    },
                    (error) => {
                        sp.fail();
                        console.log(error);
                    }
                );
        }
        return jsonTasks;
    }
    
    getListOfTaskObjects(sectionsWithTasks) {
        let tasks  = [];
        for (const tasksForSection of sectionsWithTasks) {
            for (const task of tasksForSection) {
                // Filter by sprint if it was chosen
                if (this.arguments.sprint) {
                    let sprintField = task.custom_fields.find((field) => field.name == 'BT Sprint - FY22');
                    if (!sprintField || !sprintField.display_value 
                            || !sprintField.display_value.toUpperCase().includes(this.arguments.sprint.toUpperCase())) 
                        {
                        continue;
                    }
                }

                let teamField = task.custom_fields.find((field) => field.name == 'Aquiva Team');
                let teamValue = teamField && teamField.display_value ? `(${teamField.display_value})` : '';
                let taskURL = `${this.baseURL}${task.gid}`;
                let taskObject = new AsanaTask(task.name, task.gid, taskURL, teamValue);
                tasks.push(taskObject);
            }
        }
    
        if (tasks.length == 0) {
            console.log('No Tasks found');
            process.exit(1)
        }
        return tasks
    }
    
    handleBranches(tasks) {
        let sp = spinner(`Handling tasks: `);
        sp.start();
        for (const task of tasks) {
            process.stdout.write('|')
            // Find remote Branches
            let taskBranches = this.runInTerminal(`git branch --remotes | grep ${task.id} | tr '\n' ' '`);
            if (!taskBranches && !this.arguments.short) {
                continue;
            }
    
            // Add Branches to object
            let taskBranchesSplit = taskBranches.trim().split(/\s+/);
            for (const branch of taskBranchesSplit) {
                task.branches.push(new Branch(branch.trim()));
            }
    
            // Check for reverts
            let revertCommits = this.runInTerminal(`git log --oneline | grep '${task.id}' | grep -i revert`);
            if(!revertCommits) {
                task.hasReverts = false;
            }
    
            // Check if fully merged
            if (task.branches.length > 0) {
                for (const branch of task.branches) {
                    let lastCommit = this.runInTerminal(`git log ${branch.name} -1 --oneline | awk '{print $1}'`);
                    // let lastCommitDateAbs = this.runInTerminal(`git log ${branch.name} -1 --pretty=format:"%ad" --date=format:'%d.%m.%Y'`);
                    let lastCommitDateAbs = this.runInTerminal(`git log ${branch.name} -1 --pretty=format:"%ad" --date=local`);
                    let lastCommitDateRel = this.runInTerminal(`git log ${branch.name} -1 --pretty=format:"%ad" --date=relative`);
                    branch.lastCommitDateAbsolute = lastCommitDateAbs;
                    branch.lastCommitDateRelative = lastCommitDateRel;
                    branch.lastCommitDateObject = new Date(lastCommitDateAbs);

                    let lastCommitIsMerged = this.runInTerminal(`git log --oneline | grep ${lastCommit}`);
                    if(!lastCommitIsMerged) {
                        branch.isMerged = false;
                    }
                }
            }
        }
        sp.succeed();
    }

    printTasks(tasks) {
        console.log('\n' + '='.repeat(120) + '\n');

        // Sort by date
        tasks.sort((a,b) => {
            if(a.branches.length == 0) {
                return -1;
            } 
            if(b.branches.length == 0) {
                return 1;
            }
            a.branches.sort((a,b) => a.lastCommitDateObject - b.lastCommitDateObject);
            b.branches.sort((a,b) => a.lastCommitDateObject - b.lastCommitDateObject);

            return a.branches[a.branches.length - 1].lastCommitDateObject - b.branches[b.branches.length - 1].lastCommitDateObject;
        });

        for (const task of tasks) {
            // Skip tasks without problems when --short
            if (this.arguments.short && task.isReady()) {
                continue;
            }
            this.printTask(task);
        }
    }
    
    printTask(task) {
        let exclamationMark = '\u2757';
        let redCross = '\u274C';
        let checkMark = '\u2705';

        if (task.hasReverts) {
            console.log(`${exclamationMark} Has Reverts`);
        }
        if (task.branches.length > 1) {
            console.log(`${exclamationMark} Found ${task.branches.length} branches`);
        }
    
        console.log(`   ${task.team}${task.name.substring(0, 70)} -> ${task.url}`);
    
        if (task.branches.length == 0) {
            console.log(`   ${task.id} -> No Branch\n\n`);
            return;
        }
        
        for (const branch of task.branches) {
            let okPrint = `${redCross} `;
            // let isMergedPrint = " -> NOT Merged";
            if (branch.isMerged) {
                okPrint = `${checkMark} `;
                // isMergedPrint = " -> Merged";
            }
            console.log(`${okPrint}${task.id} -> ${branch.name} -> ${branch.getFormattedDate()} (${branch.lastCommitDateRelative})`)
        }
        console.log("\n");
    }
}

class Branch {
    name;
    isMerged;
    lastCommitDateObject;
    lastCommitDateAbsolute;
    lastCommitDateRelative;
    lastCommitDateFormatted;

    constructor(name) {
        this.name = name;
        this.isMerged = true;
    }

    getFormattedDate() {
        return `${this.lastCommitDateObject.getDate()}.${this.lastCommitDateObject.getMonth() + 1}.${this.lastCommitDateObject.getFullYear()}`
    }
}

class Section {
    name;
    id;

    constructor(name, id) {
        this.name = name;
        this.id = id;
    }
}

class AsanaTask {
    name;
    id;
    url;
    team;
    branches;
    hasReverts;

    constructor(name, id, url, team) {
        this.name = name;
        this.id = id;
        this.url = url;
        this.team = team;
        this.branches = [];
        this.hasReverts = undefined;
    }

    isReady() {
        return this.branches.length == 1 && this.branches[0].isMerged && !this.hasReverts
    }
}

// Prorgam run


// let asd = new Date('Tue Mar 21 11:13:02 2023');
// console.log(asd.getDate());
// console.log(asd.getMonth() + 1);
// console.log(asd.getFullYear());
const releaseValidator = new ReleaseValidator();
releaseValidator.run();
